### PR TITLE
devops: task to run coverage locally

### PR DESCRIPTION
**Pull request by max-sixty**
I couldn't get the VSCode extension to show coverage, but may return to this.

### CODE REVIEW

File: Taskfile.yml [hunk 1]
```diff
@@ -318,6 +318,25 @@ tasks:
...
+      # https://github.com/mozilla/grcov#generate-a-coverage-report-from-coverage-artifacts
+      CARGO_INCREMENTAL: 0
+      RUSTFLAGS: "-Cinstrument-coverage"
+      LLVM_PROFILE_FILE: "cargo-test-%p-%m.profraw"
+      # Use a different target dir so we don't poison the cache
+      CARGO_TARGET_DIR: "./target-cov"
+    cmds:
+      # Setup
+      - cargo install grcov
+      - rustup component add llvm-tools-preview
```
Comment (aljazerzen) [inline Taskfile.yml]: I'd ask for these setup tasks not be present here, as this does not work on my machine.

**Review by aljazerzen — COMMENTED** [review]
I remember using [tarpaulin](https://github.com/xd009642/tarpaulin) and it was quite pleasant.

**Comment by aljazerzen** [comment]
I've pushed [coverage-llvm-cov](https://github.com/PRQL/prql/compare/main...coverage-llvm-cov), it adapts this PR and emits lcov.info file, which can be read by [a few different vscode extensions](https://marketplace.visualstudio.com/items?itemName=ryanluker.vscode-coverage-gutters).

**Comment by max-sixty** [comment]
> I remember using [tarpaulin](https://github.com/xd009642/tarpaulin) and it was quite pleasant.

`-C instrument-coverage` & `grcov` additionally tracks the lines hit by calling `prqlc` in tests, which is much much better, since about half of our gaps weren't actually gaps. (This was done for Ci in https://github.com/PRQL/prql/pull/3987 though spreads across a few PRs so maybe wasn't easy to track)

**Comment by max-sixty** [comment]
> I've pushed [coverage-llvm-cov](https://github.com/PRQL/prql/compare/main...coverage-llvm-cov),

Ah this seems better than using `-C instrument-coverage` & `grcov` alone; seems like it's a wrapper of that.

Annoying that I hadn't seen this before! 

Probably we should switch CI into using that too...

**Comment by max-sixty** [comment]
This is changed to use `coverage-llvm-cov`! It works with VS Code's coverage gutters.

I'll plan to merge unless there are any objections

**Comment by max-sixty** [comment]
@aljazerzen I didn't put this in the main test task:
- It's much slower — 54s vs 20s when running from an existing cache — partly because it has to recompile the existing binaries, because incremental caching is disabled
- It would not be cached with other builds

**Comment by aljazerzen** [comment]
> incremental caching is disabled

I don't think rust supports incremental compilation at all - you mean it has to re-compile all of the dependencies? 

In my commit, I've set:
```
      # Use a different target dir so we don't poison the cache
      CARGO_LLVM_COV_TARGET_DIR: ../target-cov
```
... so first run takes ~1min and re-run takes ~20sec, even if I compile something else inbetween.

Are we using this task in the pipelines too?

**Comment by max-sixty** [comment]
> I don't think rust supports incremental compilation at all - you mean it has to re-compile all of the dependencies?

https://blog.rust-lang.org/2016/09/08/incremental.html 

On my commit, it seems to be recompiling our own binaries on each run...

---

On your commit I get

```sh
task prqlc:test

error: failed to parse nextest config at `/Users/maximilian/workspace/prql/.config/nextest.toml`

Caused by:
  invalid type: map, expected a string for key `profile.default.slow-timeout`
error: process didn't exit successfully: `/Users/maximilian/.rustup/toolchains/1.74.0-aarch64-apple-darwin/bin/cargo nextest run --manifest-path /Users/maximilian/workspace/prql/Cargo.toml --target-dir /Users/maxim
ilian/workspace/prql/prqlc/../target-cov --features=test-dbs -p prqlc-ast -p prqlc-parser -p prql-compiler -p prqlc -p prql-compiler-macros -p compile-files -p prql -p prql-java -p prql-js -p prqlc-clib -p prql-pyt
hon` (exit status: 96)

**Comment by aljazerzen** [comment]
> https://blog.rust-lang.org/2016/09/08/incremental.html

I was aware of experimental support for it, but I didn't know it's rolled out yet!
I just checked and does work very well! It is enabled by default for dev and test profiles, so you are correct that we'd not be taking advantage of that here.

---

But regardless of incremental compilation, when there was no changes to source, no crates should be rebuilt.

Ah, this is why:

> cargo llvm-cov/cargo llvm-cov run/cargo llvm-cov nextest cleans some build artifacts by default to avoid false positives/false negatives due to old build artifacts. This behavior is disabled when --no-clean

(from [here](https://github.com/taiki-e/cargo-llvm-cov#basic-usage))

---

About nextest config error: maybe try updating nextest? It works on my computer^tm with nextest 0.9.54.

**Comment by max-sixty** [comment]
> About nextest config error: maybe try updating nextest? It works on my computer^tm with nextest 0.9.54.

Sorry, you're right!

---

OK, I'll merge this, then!

**Pull request merged**